react/week3/ToDoList.js

```diff
@@ -0,0 +1,165 @@
+import React, { useEffect, useState } from "react";
+import PropTypes from "prop-types";
+// styling the border
+const TodoComponent = (props) => {
+  return (
+    <div
+      style={{
+        width: "700px",
+        marginLeft: "300px",
+        fontSize:"20px",
+        fontFamily:"cursive",
+      }}
+    >
+      {props.children}
+    </div>
+  );
+};
+const API_URL =
+  "https://gist.githubusercontent.com/benna100/391eee7a119b50bd2c5960ab51622532/raw";
+function TodoList({ id, description, deadline, removeTodo }) {
+  const [isDone, setIsDone] = useState(false);
+  const [edit, setEdit] = useState(false);
+  const [newDescription, setNewDescription] = useState(description);
+  const updateTodo = () => {
+    setEdit(!edit);
+  };
+  return (
+    <>
+      <TodoComponent>
+        <div
+          className="list-of-todos"
+          style={
+            isDone ? { textDecoration: "line-through" } : { textDecoration: "" }
+          }
+        >
+          {edit ? (
+            <input
+              type="text"
+              value={newDescription}
+              onChange={(event) => {
+                const value = event.target.value;
+                setNewDescription(value);
+              }}
+            />
+          ) : (
+            newDescription
+          )}{" "}
+          | {deadline}
+          <div className="buttons">
+            <input type="checkbox" onClick={() => setIsDone(!isDone)} />
+            <button className="delet-button" onClick={() => removeTodo(id)}>
+              Delete
+            </button>
+            <button className="edit-button" onClick={updateTodo}>
+              {edit ? "Update" : "Edit"}
+            </button>
+          </div>
+        </div>
+      </TodoComponent>
+    </>
+  );
+}
+//component is for input elements
+function RenderTodos() {
+  const [description, setDescription] = useState("");
+  const [deadline, setDeadline] = useState("");
+  const [todos, setTodos] = useState([]);
+
+  const minimumDate = () => {
+    let today = new Date();
+    let dd = today.getDate();
+    let mm = today.getMonth() + 1; 
+    let yyyy = today.getFullYear();
+    if (dd < 10) {
+      dd = "0" + dd;
+    }
+    if (mm < 10) {
+      mm = "0" + mm;
+    }
+    today = yyyy + "-" + mm + "-" + dd;
+    return today;
+  };
+  // fetching the data from api
+  useEffect(() => {
+    fetch(API_URL)
+      .then((response) => response.json())
+      .then((data) => {
+        setTodos(data);
+      });
+  }, []);
+  //Add to the todos
+  const addTodos = (event) => {
+    event.preventDefault();
+    if (description !== "") {
+      const newTodos = {
+        id: todos.length + 1,
```

Comment:
This is not the best way to construct a unique ID. 
Let me show you an example:
You have an array of todos ---> `[todo1, todo2, todo3]`
Let's say the user adds a new todo `todo4`, `todo4` will have the id of todos.length + 1 = 4.
Then the user realizes they don't need `todo1` so they remove that from the list.
Now you have `[todo2, todo3, todo4]` 
If the user adds a new todo, `todo5`, the id will be todos.length + 1 = 4 again. So you end up with 2 items that have the same id.

Here you can find some unique ID alternatives:
https://dev.to/rahmanfadhil/how-to-generate-unique-id-in-javascript-1b13